tests/statedb_insert_chain_transaction_test.go

```diff
@@ -0,0 +1,766 @@
+package tests
+
+import (
+	"context"
+	"crypto/ecdsa"
+	"fmt"
+	"math/big"
+	"testing"
+
+	"github.com/ledgerwatch/turbo-geth/accounts/abi/bind"
+	"github.com/ledgerwatch/turbo-geth/accounts/abi/bind/backends"
+	"github.com/ledgerwatch/turbo-geth/common"
+	"github.com/ledgerwatch/turbo-geth/consensus/ethash"
+	"github.com/ledgerwatch/turbo-geth/core"
+	"github.com/ledgerwatch/turbo-geth/core/types"
+	"github.com/ledgerwatch/turbo-geth/core/vm"
+	"github.com/ledgerwatch/turbo-geth/crypto"
+	"github.com/ledgerwatch/turbo-geth/ethdb"
+	"github.com/ledgerwatch/turbo-geth/params"
+	"github.com/ledgerwatch/turbo-geth/tests/contracts"
+)
+
+func TestInsertIncorrectStateRootDifferentAccounts(t *testing.T) {
+	data := getGenesis()
+	from := data.addresses[0]
+	fromKey := data.keys[0]
+	to := common.Address{1}
+
+	blockchain, blocks, receipts, err := genBlocks(data.genesisSpec, map[int]tx{
+		0: {
+			getBlockTx(from, to, big.NewInt(1000)),
+			fromKey,
+		},
+		1: {
+			getBlockTx(from, to, big.NewInt(2000)),
+			fromKey,
+		},
+	})
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// BLOCK 1
+	incorrectHeader := blocks[0].Header()
+	incorrectHeader.Root = blocks[1].Header().Root
+
+	if blocks[0].Header().Root == incorrectHeader.Root {
+		t.Fatal("roots are the same")
+	}
+
+	incorrectBlock := types.NewBlock(incorrectHeader, blocks[0].Transactions(), blocks[0].Uncles(), receipts[0])
+	if _, err = blockchain.InsertChain(types.Blocks{incorrectBlock}); err == nil {
+		t.Fatal("should fail")
+	}
+
+	// insert a correct block
+	blockchain, blocks, _, err = genBlocks(data.genesisSpec, map[int]tx{
+		0: {
+			getBlockTx(data.addresses[1], to, big.NewInt(5000)),
+			data.keys[1],
+		},
+	})
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	if _, err = blockchain.InsertChain(blocks); err != nil {
+		t.Fatal(err)
+	}
+
+	st, _, _ := blockchain.State()
+	if !st.Exist(to) {
+		t.Error("expected account to exist")
+	}
+
+	if balance := st.GetBalance(from); balance.Cmp(big.NewInt(1000000000)) != 0 {
```

Comment:
I think you should use a more precise check. Like, check with some delta.

Reply:
We check that balance is the same after incorrect block.